### PR TITLE
feat: use mempool for sweep fees (part 3)

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

Closes: #758 

## Changes

Swaps out the outstanding sweep transaction implementation from using the database to using the mempool via the new methods introduced in part 1 and 2.

I think that we should be able to remove sweep tables after this, but I didn't want to address that here.

## Testing Information

The "big" test for this was updated to use Bitcoin core.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/transaction_coordinator.rs [hunk 1]
```diff
@@ -1360,6 +1350,110 @@ where
...
     }
+
+    /// Assesses the total fees paid for any outstanding sweep transactions in
+    /// the mempool which may need to be RBF'd. If there are no sweep
+    /// transactions which are spending the signer's UTXO, then this function
+    /// will return [`None`].
+    ///
+    /// TODO: This method currently blindly assumes that the mempool transactions
+    /// are correct. Maybe we need some validation?
+    #[tracing::instrument(skip_all, fields(%chain_tip, signer_utxo = %signer_utxo.outpoint))]
```
Comment (djordon) [inline signer/src/transaction_coordinator.rs:1361]: We probably want to remove the `chain_tip` from here.

**Review by djordon — APPROVED** [review]
Awesome, looks good.

I had some small comments.

File: signer/src/transaction_coordinator.rs [hunk 2]
```diff
@@ -1360,6 +1350,110 @@ where
...
+
+        // Retrieve fees for all descendant transactions. If there were no
+        // descendants then this will just result in an empty list.
+        let descendant_fees = try_join_all(descendant_txids.iter().map(|txid| {
+            let bitcoin_client = bitcoin_client.clone();
+            async move {
+                bitcoin_client
+                    .get_transaction_fee(txid, Some(TransactionLookupHint::Mempool))
+                    .await
+            }
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1435]: nah it yells at you if you don't first clone it outside of the async closure (tried it again now)

File: signer/src/transaction_coordinator.rs (hunk 1 above)
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs:1361]: Actually wasn't needed at all, must have forgotten to remove it at some point

File: signer/src/transaction_coordinator.rs [hunk 3]
```diff
@@ -1360,6 +1350,110 @@ where
...
+        .into_iter()
+        .max_by_key(|(_, fees)| fees.fee);
+
+        // Since we got the transaction ids from bitcoin-core, these should
+        // not be missing, but we double-check here just in case (it could
+        // happen that the client has failed-over to the next node which isn't
+        // in sync with the previous one, for example).
+        let Some((best_sweep_root_txid, fees)) = best_sweep_root else {
+            tracing::warn!(
+                utxo_outpoint = %signer_utxo.outpoint,
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs]: [cc942a0](https://github.com/stacks-network/sbtc/pull/881/commits/cc942a05e0fe1be2d2bf70cf87cd2253685e803d)

File: signer/src/transaction_coordinator.rs [hunk 4]
```diff
@@ -1360,6 +1350,110 @@ where
...
+        // If we have some transactions, we need to find the one that pays the
+        // highest fee. This is the transaction that we will use as the root of
+        // the sweep package. Note that even if only one transaction was
+        // returned above, we still need to get the fee for it, which is why
+        // there's no special logic for one vs multiple.
+        //
+        // This can technically error if the mempool transactions are not found,
+        // but it shouldn't happen since we got the transaction ids from
+        // bitcoin-core itself.
+        let best_sweep_root = try_join_all(mempool_txs_spending_utxo.iter().map(|tx| {
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/transaction_coordinator.rs]: [cc942a0](https://github.com/stacks-network/sbtc/pull/881/commits/cc942a05e0fe1be2d2bf70cf87cd2253685e803d)

**Pull request merged**